src/lib/utils/format_duration.cpp

```diff
@@ -1,53 +1,43 @@
 #include "format_duration.hpp"
 
-#include <cmath>
 #include <sstream>
 
+#include <boost/hana/for_each.hpp>
+#include <boost/hana/tuple.hpp>
+
 #include "utils/assert.hpp"
 
 namespace opossum {
 
 std::string format_duration(const std::chrono::nanoseconds& total_nanoseconds) {
-  // Returns the result of a positive integer division rounded both to zero and to the nearest int so that 60900ms
-  // are returned as 61s, not 60s
-  const auto div_both = [](const auto a, const auto b) {
-    DebugAssert(a >= 0 && b > 0, "Can only operate on positive integers");
-    const auto round_to_zero = a / b;
-    auto round_to_nearest = a / b;
-    if (a - round_to_zero * b >= b / 2) round_to_nearest += 1;
-    return std::make_pair(round_to_zero, round_to_nearest);
-  };
-
-  auto nanoseconds_remaining = total_nanoseconds.count();
-
-  const auto minutes = div_both(nanoseconds_remaining, 60'000'000'000);
-  nanoseconds_remaining -= minutes.first * 60'000'000'000;
-
-  const auto seconds = div_both(nanoseconds_remaining, 1'000'000'000);
-  nanoseconds_remaining -= seconds.first * 1'000'000'000;
-
-  const auto milliseconds = div_both(nanoseconds_remaining, 1'000'000);
-  nanoseconds_remaining -= milliseconds.first * 1'000'000;
-
-  const auto microseconds = div_both(nanoseconds_remaining, 1'000);
-  nanoseconds_remaining -= microseconds.first * 1'000;
-
-  const auto nanoseconds = nanoseconds_remaining;
+  constexpr auto TIME_UNIT_ORDER =
+      boost::hana::to_tuple(boost::hana::tuple_t<std::chrono::minutes, std::chrono::seconds, std::chrono::milliseconds,
+                                                 std::chrono::microseconds, std::chrono::nanoseconds>);
+  const std::vector<std::string> unit_strings = {" min", " s", " ms", " µs", " ns"};
+
+  std::chrono::nanoseconds remaining_nanoseconds = total_nanoseconds;
+  std::vector<uint64_t> floor_durations;
+  std::vector<uint64_t> round_durations;
+  boost::hana::for_each(TIME_UNIT_ORDER, [&](const auto duration_t) {
+    using DurationType = typename decltype(duration_t)::type;
+    auto floored_duration = std::chrono::floor<DurationType>(total_nanoseconds);
+    floor_durations.emplace_back(floored_duration.count());
+    round_durations.emplace_back(std::chrono::round<DurationType>(remaining_nanoseconds).count());
+    remaining_nanoseconds -= floored_duration;
+  });
```

Comment:
I somehow feel that there should be another more general solution to that without using `boost::hana`, we cannot be the first to solve that?

Reply:
A quick web search did not provide something better, but I guess there should be a better way

Reply:
Hm, looks like this is not as trivial either

Reply:
I found a solution without using `boost::hana`, yet I am not sure whether it is nicer.

```c++
std::string format_duration(const std::chrono::nanoseconds& total_nanoseconds) {
  const auto unit_strings = std::vector<std::string>{" min", " s", " ms", " µs", " ns"};
  bool finished = false;
  auto unit_iterator = size_t{0};
  std::stringstream stream;

  const auto iterate_time_units = [&](auto&&... time_unit_pairs) {
    const auto try_write_output = [&](auto& time_unit_pair) {
      if (finished) return;
      using LargerDuration = typename std::decay_t<decltype(time_unit_pair.first)>;
      using SmallerDuration = typename std::decay_t<decltype(time_unit_pair.second)>;
      const auto floored_duration = std::chrono::floor<LargerDuration>(total_nanoseconds);
      const auto is_last_unit = unit_iterator == unit_strings.size() - 1;
      if (floored_duration > 0ns || is_last_unit) {
        finished = true;
        stream << floored_duration.count() << unit_strings.at(unit_iterator);
        if (!is_last_unit) {
          const auto rounded_duration = std::chrono::round<SmallerDuration>(total_nanoseconds - floored_duration);
          stream << " " << rounded_duration.count() << unit_strings.at(unit_iterator + 1);
        }
      }
      ++unit_iterator;
    };

    (try_write_output(time_unit_pairs), ...);
  };

  iterate_time_units(std::make_pair(0min, 0s), std::make_pair(0s, 0ms), std::make_pair(0ms, 0us),
                     std::make_pair(0us, 0ns), std::make_pair(0ns, 0ns));

  return stream.str();
}
```

Reply:
Not really :D